tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+
+      depFile.deps = this._scanDeps(absImportedPath, depFile.data);
+    });
+
+    return deps;
+  }
+
+  // Returns a { installPath, inNodeModules } record indicating where to
+  // install the given file via meteorInstall, and whether it resides in a
+  // node_modules directory. May return undefined if the file should not
+  // be installed on the current architecture.
+  _getInstallInfo(absPath) {
+    const info =
+      this._getNodeModulesInstallInfo(absPath) ||
+      this._getSourceRootInstallInfo(absPath);
+
+    if (! info) {
+      return;
+    }
+
+    if (this.name) {
+      // If we're bundling a package, prefix info.path with
+      // node_modules/<package name>/.
+      info.path = pathJoin("node_modules", this.name, info.path);
+    } else {
+      // If we're bundling an app, prefix info.path with app/.
+      info.path = pathJoin("app", info.path);
+    }
+
+    // Note that info.inNodeModules may be false even if info.path now
+    // contains a node_modules directory.
+    return info;
+  }
+
+  _getNodeModulesInstallInfo(absPath) {
+    if (this.nodeModulesPath) {
+      const relPathWithinNodeModules =
+        pathRelative(this.nodeModulesPath, absPath);
+
+      if (relPathWithinNodeModules.startsWith("..")) {
+        // absPath is not a subdirectory of this.nodeModulesPath.
+        return;
+      }
+
+      if (! this._hasKnownExtension(relPathWithinNodeModules)) {
+        // Only accept files within node_modules directories if they
+        // have one of the known extensions.
+        return;
+      }
+
+      return {
+        // Install the module into the local node_modules directory within
+        // this app or package.
+        path: pathJoin("node_modules", relPathWithinNodeModules),
+        // The original path was contained by a node_modules directory.
+        inNodeModules: true,
+      };
+    }
+  }
+
+  _getSourceRootInstallInfo(absPath) {
+    const installPath = pathRelative(this.sourceRoot, absPath);
+
+    if (installPath.startsWith("..")) {
+      // absPath is not a subdirectory of this.sourceRoot.
+      return;
+    }
+
+    const dirs = this._splitPath(pathDirname(installPath));
+    const bundlingClientApp =
+      ! this.name && // Indicates we are bundling an app.
+      archMatches(this.bundleArch, "web");
+
+    for (let dir of dirs) {
+      if (dir.charAt(0) === "." ||
+          dir === "packages" ||
+          dir === "programs" ||
+          dir === "cordova-build-override") {
+        // These directories are never loaded as part of an app.
+        return;
+      }
+
+      if (bundlingClientApp && (dir === "server" ||
+                                dir === "private")) {
+        // If we're bundling an app for a client architecture, any files
+        // contained by a server-only directory that is not contained by
+        // a node_modules directory must be ignored.
+        return;
+      }
+
+      if (dir === "node_modules") {
```

Comment:
You rarely see `node_modules` in an imported/required module identifier, because the way to import something from a `node_modules` directory is simply to `require("bar")` *without* a `./` or `../`.

This whole loop is attempting to determine whether a module should be allowed to be installed. This clause specifically relaxes all the restrictions above for files in `node_modules` directories, because you wouldn't want an npm package to break because it put some of its modules in a `server/` or `packages/` directory.